### ISSUE TITLE
2-site iPEPS AD: joint metric + retraction needed for general (non-Heisenberg) models

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#328]
## Summary

The 2-site iPEPS AD path is unstable for **any model** when `gs_c4v=False` because the metric preconditioner in `src/tenax/algorithms/_metric_precond.py` only handles the **per-site local metric** (`precondition_gradient_multisite` is explicitly block-diagonal in site index). Joint flat directions across the (A, B) parameter pair — e.g. `A → λA, B → (1/λ)B` — pass through the block-diagonal preconditioner unimpeded, and L-BFGS follows them into non-variational CTM regions.

Today's benches (2026-04-15, 2-site C4v Heisenberg D=2) confirm the pathology:

| gs_c4v | AD mode | LS | χ | result |
|---|---|---|---|---|
| True | explicit | Armijo | 16 | ✓ E_best = -0.6590 |
| False | explicit | Armijo | 8 | ✗ E = -0.9464 (below QMC -0.6694) |
| True | implicit | Armijo | 16 | ✓ E_best = -0.6388 |
| False | implicit | Armijo | 16 | 💥 E_best = -33.85 |

The current `gs_c4v=True` escape hatch is **AFM Heisenberg / bipartite / spin-½ specific** — it hard-codes `U_sub = iσ_y` (Marshall sublattice rotation) at `ipeps_optimize.py:1291` and raises for `d_loc ≠ 2`. This is not a general solution for:

- Higher spin (S=1, S=2, ...)
- Frustrated magnets (triangular, kagome)
- Fermionic models (Hubbard, t-J)
- Anisotropic / chiral / Kitaev-like models
- Any model without a Marshall sign structure

PR #327 closes the immediate test failure (#325) by switching the test to `gs_c4v=True`, but that masks rather than fixes the underlying instability for general models.

## Root cause

`_metric_precond.py` implements **metric preconditioning** (QGT natural gradient via GMRES on `(N + δI) g' = g`), not **Riemannian retraction**. Two things are missing for stable general 2-site optimization:

### 1. Joint multi-site metric
`precondition_gradient_multisite` (line 174) explicitly comments:

```
The metric is block-diagonal in site index (local approximation),
so each site's gradient is preconditioned using its own environment.
```

This handles per-site rescaling `A → λA`, but joint flat directions like `A → λA, B → (1/λ)B` and joint gauge modes that couple A and B through the shared CTM environment are **not in the per-site kernel** and pass through the preconditioner. L-BFGS is then free to follow them.

### 2. No retraction step
There's no projection of L-BFGS iterates back onto a manifold (unit sphere, Stiefel, etc.) after each update. Tensors are only normalized inside the loss function (`A → A/||A||`), but the raw `params` that L-BFGS carries in its history are unconstrained. Curvature pairs `(s_k, y_k)` accumulate across iterates that live in the flat direction, corrupting the Hessian approximation.

## Proposed fix (to start in a follow-up session)

Three options, ordered by completeness:

### Option A — minimal: explicit retraction
Project L-BFGS iterates onto `||A|| = ||B|| = 1` after each accepted step. ~10 lines in `_optimize_gs_ad_tensor_2site`. Eliminates the joint norm-rescaling flat direction at the parameter level without touching the metric. Fast to implement; doesn't require new math.

### Option B — joint multi-site metric
Implement `norm_environment_matvec_joint(A, B, env_A, env_B, v_A, v_B)` that computes the full 2×2 block QGT across the (A, B) pair, not just block-diagonal. Then extend `precondition_gradient_multisite` to call the joint metric when the config requests it. Costs ~4× per-matvec relative to block-diagonal (two cross-blocks plus the two diagonals), but correctly projects out joint flat directions. Needs a new CTM environment contraction (the "double-site" norm environment).

### Option C — full Riemannian L-BFGS
Combine (A) and (B): retraction onto the unit manifold + joint QGT as the natural-gradient preconditioner + L-BFGS history stored in tangent space with parallel transport between iterates. Strictly best but significantly more code. Use JAX's `optax.projected_gradient` or `riemannianjax` as a starting point.

**Recommendation**: start with Option A. It's cheap, unblocks the general case today, and the data would tell us whether the joint flat direction alone is enough to cure the drift or whether the full joint metric (Option B) is also needed.

## Tests to add

- `test_ad_2site_general_noc4v_is_variational` — 2-site AD on a non-Heisenberg model (e.g. XXZ with Jz/Jxy ≠ 1, or S=1 Heisenberg), `gs_c4v=False`, assert `E_gs > physical_lower_bound`.
- `test_ad_2site_no_joint_drift` — instrument the optimizer to track `||A||` and `||B||` across steps; assert that neither diverges from the initial value by more than some tolerance (e.g. factor of 10×).

## Related

- Memory: `project_ctm_ad_bench_2026_04_15.md`, `project_multisite_ad_instability.md`, `project_2site_c4v_benchmark.md`, `project_multisite_ad_symmetry_constraint.md`
- Code: `src/tenax/algorithms/_metric_precond.py`, `src/tenax/algorithms/ipeps_optimize.py::_optimize_gs_ad_tensor_2site`
- Closed-today: #325, #327 (test workaround — not a real fix)
- Still open and relevant: `gs_c4v=True` for 2-site is only valid for spin-½ AFM Heisenberg bipartite

**Comment by yingjerkao** [comment]
Resolved via PR #332 — ships a documentation + regression-test resolution rather than a new metric or retraction.

**Why not Plan 1 (retraction) or Plan 2 (joint metric):**

1. **Plan 1 Task 8** (per-site retraction) is a no-op on post-#330 main — `_normalize_params` already runs on every accepted step (lines 254/1647/1728) and PR #330 already tangent-projects the direction.
2. **Plan 2 joint-metric** would require ~80 new lines to attack a flat direction that per-site normalization already kills at the parameterization level (joint A→λA, B→(1/λ)B leaves the product sphere unless |λ|=1).
3. Empirically, **more explicit CTM sweeps → worse drift** (monotone across steps ∈ {10, 30, 60}), pointing to CTM finite-χ pathology, not a metric/curvature issue.

**Bench matrix driving the resolution** (D=2 Heisenberg, χ=16, 30 AD steps, Armijo L-BFGS, CPU):

| path | config | E_gs |
|---|---|---|
| explicit | gs_explicit_ad_steps=10 | -1.1625 |
| explicit | gs_explicit_ad_steps=30 | -0.2576 |
| explicit | gs_explicit_ad_steps=60 | -1.1764 |
| explicit | sigma gauge, χ=8 | -2.183 |
| explicit | sigma gauge, χ=16 | -2.133 |
| **implicit** | **default** | **-0.566** |

Physical (Sandvik QMC): −0.6694.

**Backend note (surprise finding):** the iPEPS CTM+AD pipeline gives catastrophically different numerics on GPU vs CPU at identical float64 precision. Same code, same config, same seed: CPU → −0.566, CUDA → −7.47. The regression test in #332 forces CPU via `jax.default_device`. This is orthogonal to #328 and deserves a separate issue — likely a CUDA vs OpenBLAS/LAPACK eigh/QR divergence in the CTM projector.

PR #332 ships:
- Upgraded warning in `_optimize_gs_ad_tensor_2site` with bench data, split into explicit (blocker) vs implicit (advisory).
- `test_2site_noc4v_implicit_ad_is_variational_issue_328` pinning implicit-AD behavior in (-0.80, -0.30) on CPU.

Leaving Plan 2 (joint metric) as an open research path tracked in memory `project_328_explicit_ad_dead_end.md`; not needed for current users.
